projects/packages/sync/src/class-lock.php

```diff
@@ -42,18 +42,37 @@ class Lock {
 	 * @return boolean True if succeeded, false otherwise.
 	 */
 	public function attempt( $name, $expiry = self::LOCK_TRANSIENT_EXPIRY ) {
-		$name        = self::LOCK_PREFIX . $name;
-		$locked_time = get_option( $name );
+
+		global $wpdb;
+		$name = self::LOCK_PREFIX . $name;
+
+		// Options API is explicitly bypassed. This is because with high concurrency request to the
+		// site multiple processes would pass checks and send concurrent data to WP.
+
+		$row         = $wpdb->get_row( $wpdb->prepare( "SELECT option_value FROM $wpdb->options WHERE option_name = %s LIMIT 1", $name ) );
+		$locked_time = false;
+		if ( is_object( $row ) ) {
+			$locked_time = $row->option_value;
+		}
 
 		if ( $locked_time ) {
 			if ( microtime( true ) < $locked_time ) {
 				return false;
+			} else {
+				// If expired delete but don't send. That will occurr in new request to avoid race conditions.
```

Comment:
I am glad you are thinking about this case. You might consider using an update options set option_value = where option_name = and option_value = {exact previous option_value}. which would be an atomic success operation for one and only one thread if there was a race happening if you are hoping to squeeze more efficiency out of this.

Reply:
Based on ongoing conversations moving ahead with retry-after but this is a great piece of information for future development. Thanks for sharing.